documentation/swagger-checklist.md

```diff
@@ -43,82 +45,236 @@ This document defines a Swagger checklist for Microsoft Azure, along with clarif
+
+- [ ] **M2046**: A parameter object *MUST* have a `"name"` property defined with a non-empty string as its value.
+
+- [ ] **M2047**: A parameter object *MUST* have an `"in"` property defined with a non-empty string. The value can be one of "query", "header", "path", "body", "formData".
+
+- [ ] **M2048**: Every operation *MUST* have a "default" status code in the "responses" section. The "default" corresponds to the error returned by the operation.
+
+- [ ] **M2049**: "externalDocs" property *MUST* have both "url" and "description" accurately provided. More details can be found [here](https://github.com/OAI/OpenAPI-Specification/blob/master/versions/2.0.md#external-documentation-object).
+
+- [ ] **M2050**: Odata parameters like "$filter" *MUST* document the properties on which the object can be filtered on  and the supported operators in the description.
+
+- [ ] **M2050**: Odata parameters like "$filter" *MUST* have an externalDocs that provides good examples of how the "$filter" parameter can be used.
+
+- [ ] **M2051**: Parameters or properties that can only accept one value *MUST* be marked as "required" and they *MUST* have an "enum" array constraint with one valid value in it.
+
+- [ ] **M2052**: Every swagger spec *MUST* have a "securityDefinitions" object accurately defined. An example can be found over [here](https://github.com/Azure/azure-rest-api-specs/blob/master/arm-analysisservices/2016-05-16/swagger/analysisservices.json#L26).
+
+- [ ] **M2053**: An operationId in the swagger spec *MUST NOT* be split by more than one underscore. Please use the "tags" array for adding extra names.
+
+- [ ] **M2054**: If the service (in Azure) only supports PATCHING tags on a Resource in the PATCH operation and has a PUT on the same resource which is named as "CreateOrUpdate", then it *MUST NOT* expose the "PATCH" operation in the swagger spec. 
+
+### SHOULD
+
+- [ ] **S2000**: If a parameter or a model property indicates that it is some "kind/type" of an artifact that can have a value from set of possible values, then it *SHOULD* have an `"enum": ["Array of possible values"]` constraint on that entity. It *MUST* also use the ["x-ms-enum"](https://github.com/Azure/azure-rest-api-specs/blob/master/documentation/swagger-extensions.md#x-ms-enum) extension. If creating an enum is not possible, then the values need to be dcoumented in description. If the number of allowed values is huge, add a link to some kind of external documentation in the description, where those values are defined.
+
+- [ ] **S2001**: If a model is nested more than one level deep then ["x-ms-client-flatten"](https://github.com/Azure/autorest/tree/master/docs/extensions#x-ms-client-flatten) extension *MUST* be applied. This is mostly applicable while defining Resource Models in Azure Resource Manager where Resource Provider specific properties are present inside the "properties" property bag. TODO: *SHOULD* or a *MUST*? Exception if the model has a polymorphic discriminator.
```

Comment:
This is not a MUST. Use of x-ms-client-flatten depends on the context (or else we would simply always apply it and there would be no need for the extension). Optional complex values that, if they are present, have multiple required properties is one example where flattening can be inappropriate.